src/analyze/analyze-dump.c

```diff
@@ -29,51 +29,114 @@ static int dump_fallback(sd_bus *bus) {
+        if (r < 0)
+                return log_error_errno(r, "Failed to call DumpByFileDescriptor: %s",
+                                       bus_error_message(&error, r));
+
+        r = sd_bus_message_read(reply, "h", &fd);
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        fflush(stdout);
+        r = copy_bytes(fd, STDOUT_FILENO, UINT64_MAX, 0);
+        if (r < 0)
+                return r;
+
+        return 1;  /* Success */
+}
+
+static int dump_patterns_fallback(sd_bus *bus, char **patterns) {
         _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
         _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL, *m = NULL;
-        _cleanup_strv_free_ char **mangled = NULL;
         const char *text;
         int r;
 
         r = bus_message_new_method_call(bus, &m, bus_systemd_mgr, "DumpUnitsMatchingPatterns");
         if (r < 0)
                 return bus_log_create_error(r);
 
-        STRV_FOREACH(pattern, patterns) {
-                char *t;
+        r = sd_bus_message_append_strv(m, patterns);
+        if (r < 0)
+                return bus_log_create_error(r);
 
-                r = unit_name_mangle_with_suffix(*pattern, NULL, UNIT_NAME_MANGLE_GLOB, ".service", &t);
-                if (r < 0)
-                        return log_error_errno(r, "Failed to mangle name: %m");
+        r = sd_bus_call(bus, m, 0, &error, &reply);
+        if (r < 0)
+                return log_error_errno(r, "Failed to call DumpUnitsMatchingPatterns: %s",
+                                       bus_error_message(&error, r));
 
-                r = strv_consume(&mangled, t);
-                if (r < 0)
-                        return log_oom();
-        }
+        r = sd_bus_message_read(reply, "s", &text);
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        fputs(text, stdout);
+        return 0;
+}
+
+static int dump_patterns(sd_bus *bus, char **patterns) {
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL, *m = NULL;
+        int fd, r;
 
-        r = sd_bus_message_append_strv(m, mangled);
+        r = bus_message_new_method_call(bus, &m, bus_systemd_mgr, "DumpUnitsMatchingPatternsByFileDescriptor");
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_message_append_strv(m, patterns);
         if (r < 0)
                 return bus_log_create_error(r);
 
         r = sd_bus_call(bus, m, 0, &error, &reply);
         if (r < 0)
-                return log_error_errno(r, "Failed to issue method call DumpUnitsMatchingPatterns: %s",
+                return log_error_errno(r, "Failed to call DumpUnitsMatchingPatternsByFileDescriptor: %s",
                                        bus_error_message(&error, r));
 
-        r = sd_bus_message_read(reply, "s", &text);
```

Comment:
maybe move everything in this function from here on to the end into its own helper call `bus_read_fd_and_output()` or so? after all dump() above has the same 9 lines at the end.